File: codility/lessons/5_prefix_sums/passing_cars.py
# https://app.codility.com/programmers/lessons/5-prefix_sums/passing_cars/

class PrefixSumArray:

    def __init__(self, values):
        self.arr = [0] * (len(values) + 1)
        for (i, value) in enumerate(values):
            self.arr[i+1] = self.arr[i] + value

    def query_range(self, first, last):
        return self.arr[last+1] - self.arr[first]



def solution(A):
    east_direction, cars_traveling_west = 0,  PrefixSumArray(A)
    passing_cars, last_car = 0, len(A)-1

    for (car_pos, direction) in enumerate(A):
        if direction == east_direction:
            passing_cars += cars_traveling_west.query_range(car_pos+1, last_car)
        if passing_cars > 1_000_000_000:
            return -1

    return passing_cars

File: codility/lessons/5_prefix_sums/passing_cars.cpp
#include <vector>
using namespace std;

class PreffixSumArray {
     vector<int> sums;

public:
    PreffixSumArray(vector<int> &arr) {
        sums.resize(arr.size() + 1, 0);
        for (size_t i = 0; i < arr.size(); i++) {
            sums[i + 1] = sums[i] + arr[i];
        }
    }

    int query(size_t start, size_t end) {
        return sums[end] - sums[start];
    }
};

int solution(vector<int> &arr) {
    PreffixSumArray prefix_sums{arr};

    int passing_cars = 0;
    int end = arr.size();

    for (int pos = 0; pos < end; pos++) {
        if (arr[pos] == 0) {
            passing_cars += prefix_sums.query(pos + 1, end);
            if (passing_cars > 1000000000 || passing_cars < 0) {
                return -1;
            }
        }
    }

    return passing_cars;
}

